text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,187 @@
+
+### Architecture
+
+As the actual tokenizer, this tokenizer will be a standolone library,
+and will be used by meilisearch like another one.
+We decided to put it in an other repository to be able to use it in meilisearch/Meilisearch and in the future core-engine.
+
+### Implementation Details
+
+We could have 2 approach of tokenization based on the script,
+the main difficulty is to handle multi-script documents
+which need to segment each part of script and tokenize them differently,
+theses 2 approach are:
+- first iterate over the whole document, splitting it by script, and after, tokenize each part with specialized lexer
+- detect the script of the start of the document, tokenize with specialized lexer while script is the same, switch lexer if the script change
+
+#### Tokenizer
+
+```rust
+use crate::token::Token;
+use crate::internal_tokenizer::InternalTokenizer;
+
+struct Tokenizer<'a> {
+    /// script specialized tokenizer, this can be switched during
+    /// document tokenization if the document contains several scripts
+    current_tokenizer: Option<Box<dyn InternalTokenizer<'a>>>,
+    /// current character index in the document
+    current_char_index: u64,
+    /// reference on the document content
+    inner: &'a str,
+}
+
+impl<'a> Tokenizer<'a> {
+    /// create a new tokenizer detecting script
+    /// and chose the specialized internal tokenizer
+    fn new(inner: &'a str) -> Self { unimplemented!() }
+}
+
+impl<'a> Iterator for Tokenizer<'a> {
+    type Item = Token<'a>;
+
+    /// return the next token calling the `next` method of the internal tokenizer,
+    /// if the internal tokeizer return None, the function try to redetect script and chose a new tokenizer,
+    /// if no iternal tokenizer is chosen, the method return None
+    fn next(&mut self) -> Option<Self::Item> { unimplemented!() }
+}
+```
+
+#### Token
+
+```rust
+/// script of a token (https://docs.rs/whatlang/0.10.0/whatlang/enum.Script.html)
+pub type Script = whatlang::Script;
+
+/// atomic item returned by `Tokenizer::next()`
+pub struct Token<'a> {
+    word: Lexeme<'a>,
+    script: Script,
+}
+
+impl<'a> Token<'a> {
+    fn txt(&self) -> &str { unimplemented!() }
+}
+
+/// Determine the type of the token and contains a `WordSlice`
+pub enum Lexeme<'a> {
+    /// the token is a word,
+    /// meaning that it should be indexed as an important part of the document
+    Word(WordSlice<'a>),
+    /// the token is a stop word,
+    /// meaning that it can be ignored to optimize size and performance or be indexed as a Word
+    StopWord(WordSlice<'a>),
+    /// the token is a separator,
+    /// meaning that it shouldn't be indexed but used to determine word proximity
+    Separator(WordSlice<'a>)
+}
+
+/// The script, the char_index and the content of the token
+pub struct WordSlice<'a> {
+    /// content of the token
+    pub word: &'a str,
+    /// index of the first character of the token in the whole document
+    pub char_index: usize,
+}
+```
+
```

Comment:
@LegendreM @Kerollmops 
## Normalization

Normalization is an issue that hasn't been discussed. Some scripts will require words to be normalized (lowercased, deunicoded) but not all. It seems natural that this task is handled by the tokenizer. Two options:
- The token has a `normalize` method that returns a Cow<str>, and that know how to perform the normalization for this token with the information it has.
- An option `normalize` is given to the `Tokenizer`, and tokens contains a `Cow<str>` instead of a `&str`.